private-identity-audit/ffi/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+    if ptr.is_null() {
+        return;
+    }
+    Box::from_raw(ptr);
+}
+
+// ------------------------------------------------------------------------
+// Prover API
+// ------------------------------------------------------------------------
+
+/// Creates a `InitialProverResults` object from a CDD claim and a seed.
+///
+/// # Safety
+/// Caller is responsible to make sure `cdd_claim` is a valid
+/// pointer to a `CddClaimData` object, and `seed` is a random
+/// 32-byte array.
+/// Caller is responsible for deallocating memory after use.
+#[no_mangle]
+pub unsafe extern "C" fn generate_proofs(
+    cdd_claims: ArrCddClaimData,
+    committed_uids: VecEncoding,
+    seed: *const u8,
+    seed_size: size_t,
+) -> *mut ProverResults {
+    assert!(!seed.is_null());
+    assert!(seed_size == 32);
+
+    let cdd_claims = cdd_claims.to_vec();
+
+    let committed_uids = CommittedUids::decode(&mut committed_uids.to_vec().as_slice());
+    if let Ok(committed_uids) = committed_uids {
+        let mut rng_seed = [0u8; 32];
+        rng_seed.copy_from_slice(slice::from_raw_parts(seed, seed_size as usize));
+
+        _generate_proofs(cdd_claims, committed_uids, rng_seed)
+    } else {
+        println!(
+            "Error in decoding committed uids: {:?}",
+            committed_uids.unwrap_err()
+        );
+        return null_mut();
+    }
+}
+
+fn _generate_proofs(
+    cdd_claims: Vec<CddClaimData>,
+    committed_uids: CommittedUids,
+    rng_seed: [u8; 32],
+) -> *mut ProverResults {
+    let mut rng = StdRng::from_seed(rng_seed);
+
+    let result = Prover::generate_proofs::<StdRng>(&cdd_claims, &committed_uids, &mut rng);
+
+    // Log the error and return.
+    match result {
+        Err(err) => {
+            println!("Error in generating proofs: {:?}", err);
+            return null_mut();
+        }
+        Ok((initial_message_vec, final_responses_vec, re_committed_uids)) => {
+            if let Some(prover_initial_messages) = MatrixEncoding::new(initial_message_vec) {
+                if let Some(prover_final_responses) = MatrixEncoding::new(final_responses_vec) {
+                    return box_alloc(ProverResults {
+                        prover_initial_messages,
+                        prover_final_responses,
+                        committed_uids: VecEncoding::new(re_committed_uids.encode()),
+                    });
+                }
+            }
+
+            return null_mut();
+        }
+    }
+}
+
+// ------------------------------------------------------------------------
+// VerifierSetGenerator API
+// ------------------------------------------------------------------------
+
+/// Creates a `VerifierSetGeneratorResults` object from a private Uuid (as
+/// a Scalar object), a minimum set size, and a seed.
+///
+/// # Safety
+/// Caller is responsible to make sure `private_unique_identifiers`
+/// is a valid pointer to a `MatrixEncoding` object, and `seed` is a random
```

Comment:
```suggestion
/// is a `MatrixEncoding` object created through `MatrixEncoding::new`,
/// and `seed` is a random
```
It's not a pointer at least from Rust's perspective. It's on the stack, however the ABI decides to deal with it, but there is a pointer *inside*.